### ISSUE TITLE
Issue with CuDNN detection

### ISSUE THREAD

**Issue by jfsantos** [Theano/Theano#3898]
I have CuDNN V3 (final release) on my machine, however Theano recently started complaining that I have a wrong version of CuDNN and should use at least "V3 final version". I guess there's a bug in CuDNN version detection somewhere.

**Comment by carriepl** [comment]
How recently? Before updating to the Theano version that gave you this error, what Theano version were you using?

**Comment by jfsantos** [comment]
CuDNN V3 currently works on 49ca2e00 (the revision I had before pulling again from master).

**Comment by lamblin** [comment]
What is the result of calling `theano.sandbox.cuda.dnn.version()`?
Can you post the full error message and the stack trace?

**Comment by jfsantos** [comment]
Sure, here it is:

```
>>> import theano.sandbox.cuda.dnn
Using gpu device 0: Quadro K620 (CNMeM is disabled)
>>> theano.sandbox.cuda.dnn.version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jfsantos/Projects/Theano/theano/sandbox/cuda/dnn.py", line 214, in version
    if not dnn_available():
  File "/home/jfsantos/Projects/Theano/theano/sandbox/cuda/dnn.py", line 102, in dnn_available
    raise RuntimeError(dnn_available.msg)
RuntimeError: You have an old release of CuDNN (or a release candidate) that isn't supported.  Please update to at least v3 final version.
```

**Comment by lamblin** [comment]
Unfortunately, that error message does not report the version that is detected, only that it is too old.
Could you edit `theano/sandbox/cuda/dnn.py` around l. 100 and add `v` to the error message? That would be helpful for the future. Or, for now, step into `pdb` and print `v`.
Thanks,

**Comment by jfsantos** [comment]
I did this and noticed it detected version 2000. I do have both CuDNN V2 and V3 installed (respectively at `/usr/local/cuda-7.0` and `/usr/local/cuda-7.5`), but I tweak my `PATH` and `LD_LIBRARY_PATH` environment variables in order to use one or the other. Maybe an automatic method for finding CuDNN was added to Theano recently?

**Comment by lamblin** [comment]
Possible, I'm not sure what changed in Theano.
Maybe you would also have to tweak `LIBRARY_PATH` and `CPATH` so the right version is also detected at compile time?

**Comment by jfsantos** [comment]
Sorry, I think this was my mistake (I had an issue with my environment variables and was appending CUDA 7.5 paths to `PATH`/`LD_LIBRARY_PATH` instead of prepending). I got it to work now. I think the real issue is that now Theano does not support CuDNN V2 and my other code (from last month) was running with V2.

**Comment by lamblin** [comment]
OK, no problem.

**Comment by ppalasek** [comment]
Hello,

just a comment related to the issue above. I updated Theano to the latest version today and it wasn't able to detect CuDNN.

See:
```
>>> import theano
Using gpu device 0: Tesla K40c (CNMeM is disabled, CuDNN not available)
>>> print theano.__version__
0.8.0rc1.dev-11ad9bb133ea6e3b073d6e9ca2cea0500a13d6f4
```

cudnn not begin available was weird, so I spent some time trying to figure out if my environment variables were set incorrectly. I found this thread and went to see whether my cudnn version is too old:
```
>>> import theano.sandbox.cuda.dnn
>>> theano.sandbox.cuda.dnn.version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/homes/pp305/work/theano_code/theano_new/lib/python2.7/site-packages/theano/sandbox/cuda/__init__.py", line 406, in dnn_version
    dnn_available.msg)
Exception: ("We can't determine the cudnn version as it is not available", "You have an old release of CuDNN (or a release candidate) that isn't supported.  Please update to at least v3 final version.")
```

The error message "We can't determine the cudnn version as it is not available" was confusing as cudnn worked fine with the previous version. I tried it back with the old Theano version and cudnn was detected then correctly:

```
>>> import theano
Using gpu device 0: Tesla K40c (CNMeM is disabled)
>>> print theano.__version__
0.7.0.dev-15c90dd36c95ea2989a2c0085399147d15a6a477
>>> import theano.sandbox.cuda.dnn
>>> print theano.sandbox.cuda.dnn.version()
(3002, 3002)
```

So the cudnn is available, but the version is too old (it should be >= 3007 as I saw from the new code). This isn't really obvious from the "CuDNN not available" message you get when importing theano. Maybe a different error message should be used?

After updating cudnn, everything works as expected:
```
>>> import theano
Using gpu device 0: Tesla K40c (CNMeM is disabled, CuDNN 4007)
```
Thank you!

Best,
Petar

**Comment by Kevinpsk** [comment]
Hi all,

Can I ask you if anyone of you know how to setup cuDNN to work in windows 10? I am using Keras toolbox on top of theano in Anaconda in windows 10 and I am trying to setup cuDNN in order to train CNN.  

At the moment, every time I train a CNN I got the same message you got previously:
"CNMeM is disabled, cuDNN not available"

I downloaded the cuDNN library and extracted the files to my CUDA7.5 folder as shown below,
![1](https://cloud.githubusercontent.com/assets/9253234/15987656/00a5ef7a-3028-11e6-8d96-1090ba703191.PNG)

Inside the cuDNN folder, it contains this,
![2](https://cloud.githubusercontent.com/assets/9253234/15987664/73d5b818-3028-11e6-8b7b-c44d4853a783.PNG)

Which further contain the following three folders,
![3](https://cloud.githubusercontent.com/assets/9253234/15987665/8f3ca120-3028-11e6-8f55-118ab1c04952.PNG)

So can you guys tell me how shall configure my PATH variable or anything else so that Keras and theano can recognize the cuDNN library that I have downloaded.

**Comment by erfannoury** [comment]
Copy the corresponding files to their respective folders in your CUDA SDK root folder.
* `cudnn64_5.dll` to `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\bin`
* `cudnn.h` to `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\include`
* `cudnn.lib` to `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.5\lib\x64`

After copying these files, remember to add these lines to your `.theanorc` file:
```
[dnn]
enabled = True
```

**Comment by Kevinpsk** [comment]
Hi Erfannoury, 

Thank you very much for your help. I have now successfully setup cuDNN now. And it gives me three times faster training performance and I am very happy. But CNMeM is still disabled. I am using a GTX 970M, shall download anything or I can configure this in the '.theanorc' file?

Cheers,
Shuokai

**Comment by nouiz** [comment]
Use the Theano flag lib.cnmem=0.8. In your .theanorc, add:

[lib]
cnmem=0.8

**Comment by Kevinpsk** [comment]
OK, thanks a lot for the help. CNMeM is available now. Cheers,

Shuokai Pan

**Comment by Kevinpsk** [comment]
Hi Nouiz,

I was able to enable cnmem just a couple of hours before with your configuration command above, but now when I train my model again I got this error:

RuntimeError: ('initCnmem: cnmemInit call failed! Reason=CNMEM_STATUS_OUT_OF_MEMORY. numdev=1\n', 'You asked to force this device and it failed. No fallback to the cpu or other gpu device.')

Does this mean my dataset went outside my GPU memory. I have 6 GB of GPU memory, and my dataset (400 MB) is way smaller than this. Setting cnmem smaller did not solve the problem, I have to delete you command above. Any idea on what might be the wrong? 

Thanks a lot again. 
Shuokai Pan
